drivers/storage/port/scsiport/scsiport.c

```diff
@@ -849,15 +872,46 @@ ScsiPortInitialize(
         (HwInitializationData->HwFindAdapter == NULL) ||
         (HwInitializationData->HwResetBus == NULL))
     {
-        return STATUS_INVALID_PARAMETER;
+        return STATUS_REVISION_MISMATCH;
     }
 
+    PSCSI_PORT_DRIVER_EXTENSION driverExtension;
+
+    // ScsiPortInitialize may be called multiple times by the same driver
+    driverExtension = IoGetDriverObjectExtension(DriverObject, HwInitializationData->HwInitialize);
+
+    if (!driverExtension)
+    {
+        Status = IoAllocateDriverObjectExtension(DriverObject,
+                                                 HwInitializationData->HwInitialize,
+                                                 sizeof(SCSI_PORT_DRIVER_EXTENSION),
+                                                 (PVOID *)&driverExtension);
+
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("Failed to allocate the driver extension! Status 0x%x\n", Status);
+            return Status;
+        }
+    }
+
+    // set up the driver extension
+    driverExtension->RegistryPath.Buffer =
+        ExAllocatePoolWithTag(PagedPool, RegistryPath->MaximumLength, TAG_SCSIPORT);
+    driverExtension->RegistryPath.MaximumLength = RegistryPath->MaximumLength;
+    RtlCopyUnicodeString(&driverExtension->RegistryPath, RegistryPath);
+
+    driverExtension->DriverObject = DriverObject;
+
     /* Set handlers */
+    DriverObject->DriverUnload = ScsiPortUnload;
```

Comment:
I know some port drivers will let the miniport have its own unload routine, and they'll save and call the pointer if present. Is scsiport one of those, or is this safe?

Reply:
I don't see anything like that mentioned in the docs.
As far as I understand, a scsiport miniport is not supposed to allocate anything in DriverEntry, thus it doesn't need any kind of Unload routine